src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -549,11 +557,24 @@ public void SendMessages(IEnumerable<RequestMessage> messages, MessageEncoderSet
             try
             {
                 helper.EncodingMessages();
-                using (var buffer = helper.EncodeMessages(cancellationToken))
+                using (var uncompressedBuffer = helper.EncodeMessages(cancellationToken))
```

Comment:
Since some messages aren't actually sent, we need to return the actually sent messages:

                using (var uncompressedBuffer = helper.EncodeMessages(cancellationToken, out var sentMessages))